source/decorated.d

```diff
@@ -50,10 +50,18 @@ mixin template decorated(string name, alias fun)
             }
         }
 
-        static if (isCallable!(typeof(impl!(__traits(getAttributes, _internal_helper_))())))
-            return impl!(__traits(getAttributes, _internal_helper_))()();
-        else
-            return impl!(__traits(getAttributes, _internal_helper_))();
+        auto call = impl!(__traits(getAttributes, _internal_helper_))();
+        static if (isCallable!(typeof(call)))
```

Comment:
If it is not callable, what is the purpose of the contrary branching (line no 64)?

Reply:
I mean, I can rename it to callee

Reply:
I'm just trying to understand the logic of the alternative branching. What if `call` is not callable, what line number 64 falls for?

Reply:
If it's not callable it's a value. Depending on how the decorator is implemented by the user it can either return a function or a value.

```d
int boo(T)(T fun) { return 4; }

@boo
mixin decorated!("fun", { });

void main()
{
	assert(fun() == 4);
}
```

I want to replicate as much as I can from Python's implementation, however, we can't just assign any value to a variable. In Python `fun` would be an `int`, in D we cannot do it that way, so the decorator will always be a function. To "simulate" that behavior, I evaluate the calls and get the last result. If the result is a value I return it otherwise it's a closure and I run it. This prevents code like `fun()()` which is not fun, annoying, and unintuitive.

Reply:
Ah, I didn't know that was the case in Python. That makes sense either way. Thanks for the explanation.